pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+		/// Message not found.
+		MessageNotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: BlockNumberFor<T>, max_weight: Weight) -> Weight {
+			Self::service_message_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Convenience method for manually processing a message.
+		///
+		/// If the execution fails, the message gets moved to the
+		/// `FailedMessageQueue` storage.
+		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::call_index(0)]
+		pub fn process_message(origin: OriginFor<T>, nonce: T::MessageNonce) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let message = MessageQueue::<T>::take(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					Ok(())
+				}
+				Err(e) => {
+					FailedMessageQueue::<T>::insert(nonce, (message.clone(), e.error));
+
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
+				}
+			}
+		}
+
+		/// Convenience method for manually processing a failed message.
+		#[pallet::weight(T::WeightInfo::process_failed_message())]
+		#[pallet::call_index(1)]
+		pub fn process_failed_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let (message, _) =
+				FailedMessageQueue::<T>::get(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					FailedMessageQueue::<T>::remove(nonce);
+
+					Ok(())
+				}
+				Err(e) => {
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
+				}
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn service_message_queue(max_weight: Weight) -> Weight {
+			let mut weight_used = Weight::zero();
+
+			let mut processed_entries = Vec::new();
+
+			for (nonce, message) in MessageQueue::<T>::iter() {
+				processed_entries.push(nonce);
+
+				let weight = match T::MessageProcessor::process(message.clone()) {
+					Ok(post_info) => {
+						Self::deposit_event(Event::MessageExecutionSuccess { nonce, message });
+
+						post_info
+							.actual_weight
+							.unwrap_or_default()
+							// Extra weight breakdown:
+							//
+							// 1 read for the message
+							// 1 write for the event
+							// 1 write for the message removal
+							.saturating_add(T::DbWeight::get().reads_writes(1, 2))
+					}
+					Err(e) => {
+						Self::deposit_event(Event::MessageExecutionFailure {
+							nonce,
+							message: message.clone(),
+							error: e.error,
+						});
+
+						FailedMessageQueue::<T>::insert(nonce, (message, e.error));
```

Comment:
I think here it makes a lot of sense to enqueue the failing messages because we're inside the `on_idle()`, and we can not respond to the errors. But in the process_*_message() extrinsic, I think we can just fail if the message is not processed. I think there is no need to enqueue messages there, because the user can acts accordingly. Actually, they would want a fast feedback of their processing message. WDYT?

Reply:
The idea of enqueueing the failed message even in extrinsics is that it takes them out of the normal queue that gets processed on idle. 

For example:
- a message gets added to the queue
- the extrinsic to process this manually is called
- the execution fails due to some router misconfiguration (message is not removed from `MessageQueue`)
- `on_idle` gets called, causing the message to be re-executed, consuming weight but still failing.